fmt/src/git.rs

```diff
@@ -34,38 +38,45 @@ impl GitHelper {
             return Ok(None);
         }
 
-        match Repository::discover(basedir) {
+        match gix::open(basedir) {
             Ok(repo) => {
-                if repo.workdir().is_none() {
+                if repo.worktree().is_none() {
+                    let message = "bare repository detected";
                     if config.ignore.is_auto() {
-                        info!("git.ignore=auto is resolved to disabled; bare repository detected");
+                        info!("git.ignore=auto is resolved to fallback; {message}");
                         Ok(None)
                     } else {
-                        InvalidConfigSnafu {
-                            message: "Git repository is bare".to_string(),
-                        }
-                        .fail()
+                        InvalidConfigSnafu { message }.fail()
                     }
                 } else {
                     info!("git.ignore=auto is resolved to enabled");
-                    Ok(Some(GitHelper { repo }))
+                    let state = State::new(Kind::Sha1);
+                    Ok(Some(GitHelper { repo, state }))
                 }
             }
             Err(err) => {
                 if config.ignore.is_auto() {
                     info!(?err, "git.ignore=auto is resolved to disabled");
                     Ok(None)
                 } else {
-                    Err(err).context(GitOpSnafu)
+                    Err(GixOpenOpSnafu {}.into_error(Box::new(err)))
                 }
             }
         }
     }
 
-    pub fn ignored(&self, path: &Path) -> Result<bool> {
+    pub fn ignored(&self, path: &Path, is_dir: bool) -> Result<bool> {
         let path = path.canonicalize().context(ResolveAbsolutePathSnafu {
             path: path.display().to_string(),
         })?;
-        self.repo.is_path_ignored(path).context(GitOpSnafu)
+        let mut stack = self
+            .repo
+            .excludes(&self.state, None, Source::default())
+            .map_err(Box::new)
+            .context(GixExcludeOpSnafu)?;
+        let result = stack
+            .at_path(&path, Some(is_dir))
```

Comment:
The `path` should be prefix-stripped with the worktree directory to become relative to it. `gix` can't do that reliably *and* with the desired performance, as only the caller knows which root the input paths are relative to.
Without that knowledge, they need to be canonicalized which is very expensive.

Reply:
I work it around with `pathdiff` and pseudo code:

```rust
let path = path.canonicalize();
let workdir = self.repo.work_dir().canonicalize();
let at_path = pathdiff::diff_paths(path, workdir);
```

It works as expected now. But perhaps we can improve it.

Reply:
When `GitHelper` is gone, improving it will be quite straightforward I think.